CHERIify the demo + make it cheribuild-friendly

heshamelmatary opened the pull request:
This PR contains:

- Commits from @podhrmic to add the cyber-physical demo files in order to build without BESSPIN env.
- CHERIfication of boot code and drivers.

This PR relies on other submodules being CHERIfied as well such as FreeRTOS-Kernel, Plus-TCP and other newly added libraries.

heshamelmatary committed:
CHERI: Create bounded capabilities with R/W permissions on Xilinx devices

heshamelmatary committed:
CHERIfy boot code

* Add capablity sections to the linker
* Setup stack, PCC, DDC caps and other CHERI-related ops in boot.S
* Intialise globals with CHERI capabilities

heshamelmatary committed:
Make port_get_current_mtime's return unsigned long

* unsigned long is XLEN size, which is the same as the mtime CSR size.
* uint32_t (and friends) need stdint.h to be included in FreeRTOSConfig.h if
we still want to use it.
* UBaseType is defined as unsigned long (XLEN).

heshamelmatary committed:
CHERI: Create a bounded capability for the PLIC

heshamelmatary committed:
CHERI: redefine ipconfigBUFFER_PADDING for CHERI

heshamelmatary committed:
link.ld: Set UNLOAD flag for various sections

heshamelmatary committed:
CHERI: Add bits to enable building for CHERI in Makefile

heshamelmatary committed:
CHERI: Add LICENSE_CHERI file and header references

podhrmic committed:
Cyberphys demo for CHERIfication

heshamelmatary committed:
Merge branch 'hmka2' into cheridemo

heshamelmatary committed:
Enable building without IIC support

heshamelmatary committed:
iic.h: priotrise including bsp.h from -I flags

If included from another demo that already has bsp.h, using <bsp.h> will make
it include the provided -I flags rather than the cwd of the files including iic.h

heshamelmatary committed:
[CHERI] Add GLOBAL permission to device drivers capabilities

The device drivers caps can later be saved by other higher stacks within
non-global per-struct pointers.

heshamelmatary committed:
Merge remote-tracking branch 'galois/develop' into cheridemo

dmzimmerman commented on FreeRTOSIPConfig.h:
```diff
@@ -42,7 +42,34 @@ connected. */
+#else
+    #if __riscv_xlen == 64
+        #define ipconfigBUFFER_PADDING    14
+    #endif
```
If `__riscv_xlen` is 32 and `__CHERI_PURE_CAPABILITY_` is undefined, the buffer padding is left undefined. Is that what we want?

heshamelmatary replied:
Yes, it’s defined by default by FreeRTOS-TCP which assumes 32-bit builds. These defines here are added because anything other than 32-bit is considered custom

heshamelmatary replied:
In FreeRTOS_IP.h:

```
/* Space left at the beginning of a network buffer storage area to store a
 * pointer back to the network buffer.  Should be a multiple of 8 to ensure 8 byte
 * alignment is maintained on architectures that require it.
 *
 * In order to get a 32-bit alignment of network packets, an offset of 2 bytes
 * would be desirable, as defined by ipconfigPACKET_FILLER_SIZE.  So the malloc'd
 * buffer will have the following contents:
 *  uint32_t pointer;   // word-aligned
 *  uchar_8 filler[6];
 *  << ETH-header >>    // half-word-aligned
 *  uchar_8 dest[6];    // start of pucEthernetBuffer
 *  uchar_8 dest[6];
 *  uchar16_t type;
 *  << IP-header >>     // word-aligned
 *  uint8_t ucVersionHeaderLength;
 *  etc
 */

    #if ( ipconfigBUFFER_PADDING != 0 )
        #define ipBUFFER_PADDING    ipconfigBUFFER_PADDING
    #else
        #define ipBUFFER_PADDING    ( 8U + ipconfigPACKET_FILLER_SIZE )
    #endif
```

heshamelmatary committed:
compartmentalization: Add a custom fault handler for j1939 compartment

brooksdavis committed:
Allow clang to be specified via the CLANG variable

Assume clang++ is $(CLANG)++ which isn't strictly true.

brooksdavis committed:
[CHERI][boot.S] Account for new mode depedent jump instructions

From CTSRD-CHERI/FreeRTOS-Demos-CHERI-RISC-V@189c918d738694e88bf2e2fabe101b68aaf1e197

jrtc27 committed:
Fix alignment issues in linker script

jrtc27 committed:
Don't create orphan captable/caprelocs sections

jrtc27 committed:
Align sections rather than padding their starts

brooksdavis committed:
Missing ';'

Fixes 03d301b1e2d46b2c4d14059f12a4b72a4c46ef4d

jrtc27 committed:
link.ld: . doesn't affect sections when memory regions are used

Note data and uncached should already have suitably-aligned input.

brooksdavis committed:
Handle p3 processors running at 25MHz

brooksdavis committed:
Zero-init stacks on entry.

brooksdavis committed:
Add TARGET_CFLAGS to be enabled for some object files

This allows one set of files to use additional CFLAGS (e.g.,
for tests in the BESSPIN suite to enable extra protections).
Enable sub-object bounds for CHERI tests.

To connect the TARGET_CFLAGS and _TARGET_CFLAGS add a line like:

test_buffer_errors_0.o: _TARGET_CFLAGS=$(TARGET_CFLAGS)

for each file you wish to compile with the added flags. In BESSPIN
we unfortunatly need to add the full path to the object file
because out-of-tree builds are hackish.
